File: src/palette.js
// Rendu profondeur → couleur. Port JavaScript de tools/palette.py : les deux doivent
// produire la même table, de sorte que l'aperçu de contrôle et le téléphone montrent
// exactement les mêmes couleurs.
//
// Les deux modes passent par une table de 256 entrées couvrant 0 → lutMaxDepth. Un
// préréglage en bandes produit une table en marches, un préréglage continu une table
// lissée : le shader est identique dans les deux cas.

export const LUT_SIZE = 256;
export const MAX_BANDS = 8; // borne du tableau d'uniformes du shader

/**
 * Indice de table pour une profondeur.
 *
 * Règle unique, partagée par le JavaScript, le Python et le shader : l'entrée i couvre
 * l'intervalle [i/256, (i+1)/256) du domaine, donc indice = floor(ratio × 256). C'est
 * exactement ce que fait un échantillonnage de texture en NEAREST, ce qui évite au
 * shader de calculer autrement que les deux autres.
 *
 * Toute autre convention se paie : `Math.round()` arrondit au-dessus alors que `round()`
 * en Python arrondit au pair le plus proche, et les deux implémentations ne tombaient
 * pas dans la même bande aux profondeurs pile sur une borne.
 */
export function lutIndex(depth, lutMax) {
  const ratio = Math.min(Math.max(depth / lutMax, 0), 1);
  return Math.min(Math.floor(ratio * LUT_SIZE), LUT_SIZE - 1);
}

export function hexToRgb(hex) {
  const text = hex.replace('#', '');
  return [0, 2, 4].map((i) => parseInt(text.slice(i, i + 2), 16) / 255);
}

export function hexToVec4(hex, alpha = 1) {
  return [...hexToRgb(hex), alpha];
}

// --- sRGB ↔ OKLab (Björn Ottosson, 2020) --------------------------------------
// Interpoler un dégradé dans cet espace évite les teintes parasites du RVB brut :
// un rouge → vert en RVB passe par un brun sale.

const srgbToLinear = (c) => (c <= 0.04045 ? c / 12.92 : ((c + 0.055) / 1.055) ** 2.4);
const linearToSrgb = (c) => {
  const v = Math.min(Math.max(c, 0), 1);
  return v <= 0.0031308 ? v * 12.92 : 1.055 * v ** (1 / 2.4) - 0.055;
};

function srgbToOklab([r, g, b]) {
  const lr = srgbToLinear(r), lg = srgbToLinear(g), lb = srgbToLinear(b);
  const l = Math.cbrt(0.4122214708 * lr + 0.5363325363 * lg + 0.0514459929 * lb);
  const m = Math.cbrt(0.2119034982 * lr + 0.6806995451 * lg + 0.1073969566 * lb);
  const s = Math.cbrt(0.0883024619 * lr + 0.2817188376 * lg + 0.6299787005 * lb);
  return [
    0.2104542553 * l + 0.793617785 * m - 0.0040720468 * s,
    1.9779984951 * l - 2.428592205 * m + 0.4505937099 * s,
    0.0259040371 * l + 0.7827717662 * m - 0.808675766 * s,
  ];
}

function oklabToSrgb([L, A, B]) {
  const l = (L + 0.3963377774 * A + 0.2158037573 * B) ** 3;
  const m = (L - 0.1055613458 * A - 0.0638541728 * B) ** 3;
  const s = (L - 0.0894841775 * A - 1.291485548 * B) ** 3;
  return [
    linearToSrgb(4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s),
    linearToSrgb(-1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s),
    linearToSrgb(-0.0041960863 * l - 0.7034186147 * m + 1.707614701 * s),
  ];
}

// --- table -------------------------------------------------------------------

/** Couleurs des bandes, en vec4, dans le même ordre que bandLimits(). */
export function bandColors(preset) {
  if (preset.mode !== 'banded') return [];
  return preset.bands.map((band) => hexToVec4(band.color));
}

/** Bornes supérieures des bandes ; `null` dans la config signifie « jusqu'au fond ». */
export function bandLimits(preset, lutMax) {
  if (preset.mode !== 'banded') return [];
  const limits = preset.bands.map((b) => (b.max_depth_m == null ? lutMax : b.max_depth_m));
  limits[limits.length - 1] = Math.max(limits[limits.length - 1], lutMax);
  return limits;
}

function interpolate(samples, xs, ys) {
  return samples.map((x) => {
    let i = 1;
    while (i < xs.length - 1 && xs[i] < x) i += 1;
    const t = (x - xs[i - 1]) / (xs[i] - xs[i - 1] || 1);
    return ys[i - 1] + Math.min(Math.max(t, 0), 1) * (ys[i] - ys[i - 1]);
  });
}

/** Table RVBA de 256 entrées, en Uint8Array de 1024 octets. */
export function buildLut(palette, presetName) {
  const preset = palette.presets[presetName ?? palette.active_preset];
  const lutMax = palette.lut_max_depth_m;
  // Chaque entrée est évaluée au centre de l'intervalle qu'elle couvre, en cohérence
  // avec lutIndex().
  const samples = Array.from({ length: LUT_SIZE }, (_, i) => ((i + 0.5) / LUT_SIZE) * lutMax);
  const out = new Uint8Array(LUT_SIZE * 4);

  let colours;
  if (preset.mode === 'banded') {
    const limits = bandLimits(preset, lutMax);
    const bands = preset.bands.map((b) => hexToRgb(b.color));
    colours = samples.map((d) => {
      const index = limits.findIndex((limit) => d <= limit);
      return bands[index < 0 ? bands.length - 1 : index];
    });
  } else {
    const stops = [...preset.stops].sort((a, b) => a.depth_m - b.depth_m);
    const depths = stops.map((s) => s.depth_m);
    const rgbs = stops.map((s) => hexToRgb(s.color));
    if ((preset.interpolation ?? 'oklab') === 'oklab') {
      const labs = rgbs.map(srgbToOklab);
      const channels = [0, 1, 2].map((c) => interpolate(samples, depths, labs.map((l) => l[c])));
      colours = samples.map((_, i) => oklabToSrgb([channels[0][i], channels[1][i], channels[2][i]]));
    } else {
      const channels = [0, 1, 2].map((c) => interpolate(samples, depths, rgbs.map((v) => v[c])));
      colours = samples.map((_, i) => [channels[0][i], channels[1][i], channels[2][i]]);
    }
  }

  colours.forEach(([r, g, b], i) => {
    out[i * 4] = Math.round(r * 255);
    out[i * 4 + 1] = Math.round(g * 255);
    out[i * 4 + 2] = Math.round(b * 255);
    out[i * 4 + 3] = 255;
  });
  return out;
}

/** Couleur CSS correspondant à une profondeur — pour le gros chiffre et les pastilles. */
export function depthColor(palette, presetName, depth) {
  const preset = palette.presets[presetName ?? palette.active_preset];
  if (!(depth > 0)) return preset.emerged_color;
  const lut = buildLut(palette, presetName);
  const index = lutIndex(depth, palette.lut_max_depth_m) * 4;
  return `rgb(${lut[index]}, ${lut[index + 1]}, ${lut[index + 2]})`;
}

/** Entrées de légende du préréglage actif, prêtes à afficher. */
export function legendEntries(palette, presetName) {
  const preset = palette.presets[presetName ?? palette.active_preset];
  const entries = [{ color: preset.emerged_color, label: 'émergé' }];
  if (preset.mode === 'banded') {
    let previous = 0;
    preset.bands.forEach((band) => {
      const limit = band.max_depth_m;
      entries.push({
        color: band.color,
        label: limit == null ? `> ${previous} m` : previous === 0 ? `< ${limit} m` : `${previous}–${limit} m`,
      });
      if (limit != null) previous = limit;
    });
  } else {
    preset.stops.forEach((stop) => entries.push({ color: stop.color, label: `${stop.depth_m} m` }));
  }
  return entries;
}

File: tools/palette.py
"""Rampe de couleurs profondeur → RVB, interpolée en OKLab.

Ce module est la référence de calcul ; l'application en porte l'équivalent en
JavaScript pour construire sa table de correspondance de 256 entrées. Les deux
doivent donner le même résultat — d'où le test de bout en bout en fin de fichier.

OKLab : Björn Ottosson, 2020. Interpoler dans cet espace évite les teintes
parasites d'un dégradé RVB brut, notamment sur la transition rouge → vert qui
passerait par un brun sale.
"""

from __future__ import annotations

import json

import numpy as np

from common import CONFIG_DIR


def load_palette() -> dict:
    with (CONFIG_DIR / "palette.json").open(encoding="utf-8") as fh:
        return json.load(fh)


def hex_to_rgb(value: str) -> tuple[float, float, float]:
    text = value.lstrip("#")
    return tuple(int(text[i : i + 2], 16) / 255.0 for i in (0, 2, 4))


# ------------------------------------------------------------------ sRGB ↔ OKLab

def srgb_to_linear(c: np.ndarray) -> np.ndarray:
    return np.where(c <= 0.04045, c / 12.92, ((c + 0.055) / 1.055) ** 2.4)


def linear_to_srgb(c: np.ndarray) -> np.ndarray:
    c = np.clip(c, 0.0, 1.0)
    return np.where(c <= 0.0031308, c * 12.92, 1.055 * c ** (1 / 2.4) - 0.055)


_LMS_FROM_LINEAR = np.array([
    [0.4122214708, 0.5363325363, 0.0514459929],
    [0.2119034982, 0.6806995451, 0.1073969566],
    [0.0883024619, 0.2817188376, 0.6299787005],
])

_LAB_FROM_LMS = np.array([
    [0.2104542553,  0.7936177850, -0.0040720468],
    [1.9779984951, -2.4285922050,  0.4505937099],
    [0.0259040371,  0.7827717662, -0.8086757660],
])

_LMS_FROM_LAB = np.array([
    [1.0,  0.3963377774,  0.2158037573],
    [1.0, -0.1055613458, -0.0638541728],
    [1.0, -0.0894841775, -1.2914855480],
])

_LINEAR_FROM_LMS = np.array([
    [ 4.0767416621, -3.3077115913,  0.2309699292],
    [-1.2684380046,  2.6097574011, -0.3413193965],
    [-0.0041960863, -0.7034186147,  1.7076147010],
])


def srgb_to_oklab(rgb: np.ndarray) -> np.ndarray:
    lms = srgb_to_linear(np.asarray(rgb, dtype=np.float64)) @ _LMS_FROM_LINEAR.T
    return np.cbrt(lms) @ _LAB_FROM_LMS.T


def oklab_to_srgb(lab: np.ndarray) -> np.ndarray:
    lms = np.asarray(lab, dtype=np.float64) @ _LMS_FROM_LAB.T
    return linear_to_srgb((lms ** 3) @ _LINEAR_FROM_LMS.T)


# ------------------------------------------------------------------------ rampe

def build_lut(palette: dict | None = None, size: int = 256) -> tuple[np.ndarray, float]:
    """Table de correspondance RVB (uint8) échantillonnée de 0 à la profondeur du dernier palier.

    Retourne (lut, depth_max). Une profondeur d se lit à l'indice
    `round(clip(d / depth_max, 0, 1) * (size - 1))`.
    """
    palette = palette or load_palette()
    stops = sorted(palette["stops"], key=lambda s: s["depth_m"])
    depths = np.array([s["depth_m"] for s in stops], dtype=np.float64)
    colours = np.array([hex_to_rgb(s["color"]) for s in stops], dtype=np.float64)
    depth_max = float(depths[-1])

    labs = srgb_to_oklab(colours)
    samples = np.linspace(0.0, depth_max, size)

    if palette.get("interpolation", "oklab") == "oklab":
        channels = [np.interp(samples, depths, labs[:, i]) for i in range(3)]
        rgb = oklab_to_srgb(np.column_stack(channels))
    else:
        channels = [np.interp(samples, depths, colours[:, i]) for i in range(3)]
        rgb = np.column_stack(channels)

    return np.rint(np.clip(rgb, 0, 1) * 255).astype(np.uint8), depth_max


def colourise(depth: np.ndarray, palette: dict | None = None) -> np.ndarray:
    """Applique la rampe à un tableau de profondeurs (m). Les NaN ressortent en noir."""
    palette = palette or load_palette()
    lut, depth_max = build_lut(palette)
    ratio = np.clip(np.nan_to_num(depth, nan=0.0) / depth_max, 0.0, 1.0)
    return lut[np.rint(ratio * (len(lut) - 1)).astype(np.int32)]


if __name__ == "__main__":
    palette = load_palette()
    lut, depth_max = build_lut(palette)
    print(f"rampe {palette['interpolation']} · 0 → {depth_max:g} m · {len(lut)} entrées\n")
    for depth in (0, 0.5, 1, 2, 3, 5, 8, 12, 18, 25, 30, 40):
        r, g, b = lut[min(int(round(depth / depth_max * (len(lut) - 1))), len(lut) - 1)]
        print(f"  {depth:5.1f} m  #{r:02x}{g:02x}{b:02x}")
